### ISSUE TITLE
TransientStackValue incorrectly used ... leads to MNU

### ISSUE THREAD

**Issue by dalehenrich** [GsDevKit/zinc#74]
Here's a snippet of the error encountered after applying the suggested fix for Issue #72:

```
a MessageNotUnderstood occurred (error 2010), a UndefinedObject does not understand  #'add:'
1 GemServer >> writeGemLogEntryFor:titled: (envId 0) @3 line 3  [GsNMethod 176062209]
2 GemServer >> logStack:titled:inTransactionDo: (envId 0) @2 line 2  [GsNMethod 176066561]
3 GemServer >> logStack:titled: (envId 0) @2 line 2 [GsNMethod 176042241]
4 GemServer >> gemServerHandleErrorException: (envId 0) @9 line 6  [GsNMethod 176062465]
5 Error >> exceptionHandlingForGemServer: (envId 0) @2 line 2 [GsNMethod 176884993]
6 GemServer >> handleGemServerException: (envId 0) @2 line 5 [GsNMethod 176050945]
7 [] in  GemServer >> gemServer:exceptionSet:beforeUnwind:ensure: (envId 0) @3 line 9 [GsNMethod 176569601]
8 ExecBlock >> on:do: (envId 0) @3 line 42  [GsNMethod 3730689]
9 [] in  GemServer >> gemServer:exceptionSet:beforeUnwind:ensure: (envId 0) @2 line 11 [GsNMethod 176681985]
10 AbstractException >> _executeHandler: (envId 0) @3 line 8 [GsNMethod 4617217]
11 AbstractException >> _signalWith: (envId 0) @1 line 1 [GsNMethod 4594689]
12 AbstractException >> signal (envId 0) @2 line 47 [GsNMethod 4587009]
13 Object >> doesNotUnderstand: (envId 0) @9 line 10 [GsNMethod 3595265]
14 Object >> _doesNotUnderstand:args:envId:reason: (envId 0) @7 line 12  [GsNMethod 3584513]
15 [] in  ZnGemServerManagingMultiThreadedServer >> socketStreamOn: (envId 0) @3 line 5  [GsNMethod 185304321] 
```

and looking at ZnGemServerManagingMultiThreadedServer>>connections:

``` Smalltalk
connections
  connections
    ifNil: [ connections := TransientStackValue value: OrderedCollection new ].
  ^ connections value
```

I see that the TransientStackValue class is intended to be used _ONLY_ for objects that are alive on the stack (temps and args) ... TransientValue is the class to use for objects that are not guaranteed to be referenced from the stack for their entire lifetime ...

**Comment by dalehenrich** [comment]
[green](https://travis-ci.org/GsDevKit/zinc/builds/57075695)

**Comment by dalehenrich** [comment]
@HeSe @JupiterJones ... this is the next generation fix for https://github.com/GsDevKit/gsApplicationTools/issues/29 but it needs some test from you guys ... I will merge the branch (issue_72) into gs_master shortly, but will wait for your feedback before closing this bug ...

**Comment by dalehenrich** [comment]
bugfix merged (PR #75) and [tagged v2.4.3.3](https://github.com/GsDevKit/zinc/releases/tag/v2.4.3.3) take it for a spin at your leisure

**Comment by dalehenrich** [comment]
The earlier bugfix was dependent upon the method `_setNoStubbing` keeping the TransientStackValue in memory and as the following script *shows*, if the process is not active, the array can be dropped from memory eventually triggering the `transient gone` error:

```Smalltalk
      [ 
      [ 
      | x |
      x := 'x'
        ->
          (TransientStackValue
            value:
              {1.
              2.
              3}).
      opts at: 'stub'
        ifPresent: [:ignored | x _setNoStubbing ].
      Smalltalk at: #'TSV' put: x ]
        value.
      [ true ] whileTrue: [ (Delay forSeconds: 1) wait ] ]
        fork.
      Processor yield.
      [ 
      (Delay forSeconds: 1) wait.
      [ 
      (Smalltalk at: #'TSV') value value ifNil: [ nil error: 'transient gone' ].
      true ]
        whileTrue: [ (Delay forSeconds: 1) wait ] ]
        fork.
      [ 
      (Delay forSeconds: 5) wait.
      nil error: 'delay expired' ] fork.
      'hello' 
```

If this holds up (and based on the fact that @HeSe hit the same `add:` error even after the TransientStackValues were *correctly* applied it should) it means that the TransientStackValue is not sufficient and a stronger technique needs to be used ...

**Comment by dalehenrich** [comment]
Yep, looks like the **TransientStackValue** doesn't quite work as *expected*, when the process has a chance to be swapped out of active status .... and for all of the GemServer processes there is at least one other process that is on a delay loop ...

**Comment by dalehenrich** [comment]
If the the [new implementation of TransientStackValue](https://github.com/GsDevKit/GsDevKit/issues/65) is successful, then the `_setNoStubbing` calls can be removed  ... the _setNoStubbing was meant to keep us from having to create an explicit reference to the TransientStackValue from session temps, but the nice compromise is that by using a SoftRerences the session temp reference will be dropped when the process is no longer active and has been gc'ed ...

**Comment by dalehenrich** [comment]
The [new implementation is not successful](https://github.com/GsDevKit/GsDevKit/issues/65#issuecomment-93100932) ... but I should be able to arrange to set nostubbing in process scheduler, active process, and then use clientData slot in active process to hold references to the TransientStackValues ... and voila they will **live forever** ... until the process itself is gone ... at least that's the plan ...

should only have to enable this when logging to transcript ...

**Comment by rjsargent** [comment]
One of our customers encountered this error under a reasonably heavily loaded system. (Note: not too loaded!)

The following implementation is presently in use there, _but may entail some voodoo_. Clarification would be appreciated before I try to submit a pull request with a fix.

connections
	| myName |
	myName := self printString asSymbol.
	^(SpTransientGlobals transientGlobalNamed: myName)
		ifNil: [SpTransientGlobals transientGlobals at: myName put: OrderedCollection new]

Potential voodoo is using the receiver's printString as its name. It is unclear whether there can be multiple instances of the receiver's class in one session.

Also, using the receiver's printString effectively limits this solution to exactly one attribute being so managed for the entire class. A better name would incorporate the role of the attribute. e.g. #ZnGemServerManagingMultiThreadedServer_connections.

**Comment by dalehenrich** [comment]
Just merged PR #106 with a possible fix for this issue ... re-open if this does not address your issue
